// bundler/requires.js
export default {
	Button: 'const Button = require( \'react-bootstrap\' ).Button;',
	CheckboxInput: 'const CheckboxInput = require( \'general/checkbox-input\' );',
	ColorPicker: 'const ColorPicker = require( \'general/color-picker\' );',
	Column: 'const Column = require( \'react-bootstrap\' ).Col;',
	Dashboard: 'const Dashboard = require( \'general/dashboard\' );',
	DensityPlot: 'const DensityPlot = require( \'learning/d3/density-plot\' );',
	DraggableList: 'const DraggableList = require( \'general/draggable-list\' );',
	FeedbackButtons: 'const FeedbackButtons = require( \'learning/feedback\' );',
	FreeTextQuestion: 'const FreeTextQuestion = require( \'learning/free-text-question\' );',
	FunctionPlot: 'const FunctionPlot = require( \'learning/function-plot\' );',
	Grid: 'const Grid = require( \'general/grid\' );',
	Input: 'const Input = require( \'general/input\' );',
	LessonSubmit: 'const LessonSubmit = require( \'general/lesson-submit\' );',
	MultipleChoiceQuestion: 'const MultipleChoiceQuestion = require( \'learning/multiple-choice-question\' );',
	Nav: 'const Nav = require( \'react-bootstrap\' ).Nav;',
	NavItem: 'const NavItem = require( \'react-bootstrap\' ).NavItem;',
	NumberInput: 'const NumberInput = require( \'general/number-input\' );',
	Panel: 'const Panel = require( \'react-bootstrap\' ).Panel;',
	RHelp: 'const RHelp = require( \'learning/rhelp\' );',
	RPlot: 'const RPlot = require( \'learning/r/plot\' );',
	RShell: 'const RShell = require( \'learning/rshell\' );',
	RTable: 'const RTable = require( \'learning/r/table\' );',
	Row: 'const Row = require( \'react-bootstrap\' ).Row;',
	SelectInput: 'const SelectInput = require( \'general/select-input\' );',
	SliderInput: 'const SliderInput = require( \'general/slider-input\' );',
	Switch: 'Switch = require( \'learning/switch\' );',
	Slider: 'const Slider = require( \'general/slider\' );',
	Tab: 'const Tab = require( \'react-bootstrap\' ).Tab;',
	Tabs:'const Tabs = require( \'react-bootstrap\' ).Tabs;',
	TeX: 'const TeX = require( \'learning/tex\' );',
	TextInput: 'const TextInput = require( \'general/text-input\' );',
	Timer: 'const Timer = require( \'general/timer\' );',
	Well: 'const Well = require( \'react-bootstrap\' ).Well;',
	VictoryAnimation: 'const VictoryAnimation = require( \'victory\' ).VictoryAnimation;',
	VictoryArea: 'const VictoryArea = require( \'victory\' ).VictoryArea;',
	VictoryBar: 'const VictoryBar = require( \'victory\' ).VictoryBar;',
	VictoryTheme: 'const VictoryTheme = require( \'victory\' ).VictoryTheme;',
	VictoryChart: 'const VictoryChart = require( \'victory\' ).VictoryChart;',
	VictoryGroup: 'const VictoryGroup = require( \'victory\' ).VictoryGroup;',
	VictoryAxis: 'const VictoryAxis = require( \'victory\' ).VictoryAxis;',
	VictoryErrorBar: 'const VictoryErrorBar = require( \'victory\' ).VictoryErrorBar;',
	VictoryLine: 'const VictoryLine = require( \'victory\' ).VictoryLine;',
	VictoryTooltip: 'const VictoryTooltip = require( \'victory\' ).VictoryTooltip;',
	VictoryVoronoiTooltip: 'const VictoryVoronoiTooltip = require( \'victory\' ).VictoryVoronoiTooltip;',
	VideoPlayer: 'const VideoPlayer = require( \'learning/video-player\' );'
};
