// learning/rshell/index.js
// MODULES //

import React, { PropTypes } from 'react';
import ReactDom from 'react-dom';
import { Button } from 'react-bootstrap';
import request from 'request';
import ace from 'brace';
import 'brace/mode/r';
import 'brace/theme/katzenmilch';


// VARIABLES //

const OPEN_CPU = 'https://public.opencpu.org';
const OPEN_CPU_POST = '/ocpu/library/base/R/identity/identity-d';


// R SHELL //

class RShell extends React.Component {
	constructor( props ) {
		super( props );

		this.state = {
			code: this.props.code,
			result: ''
		};

		this.handleClick = () => {
			request.post( OPEN_CPU + OPEN_CPU_POST, {
				form: {
					x: this.state.code
				}
			}, ( error, response, body ) => {
				const arr = body.split( '\n' );
				if ( !error ) {
					request.get( OPEN_CPU + arr[ 1 ], ( err, getResponse, getBody ) => {
						this.setState({ result: getBody });
						this.props.onResult( this.state.result );
					});
				}
			});
		};
	}

	componentDidMount() {
		this.editor = ace.edit( ReactDom.findDOMNode( this ).firstChild );
		this.editor.getSession().setMode( 'ace/mode/r' );
		this.editor.setTheme( 'ace/theme/katzenmilch' );
		this.editor.$blockScrolling = Infinity;
		this.editor.setValue( this.state.code );
		this.editor.setOptions({
			maxLines: this.props.lines,
			minLines: this.props.lines,
		});
		this.editor.getSession().setUseWrapMode( true );
	}

	componentDidUpdate() {
		this.editor.setValue( this.state.code );
	}

	render() {
		return (
			<div className="RShell" style={{ padding: '5px', border: '2px solid #73AD21' }}>
				<div id="ace"></div>
				<Button
					bsStyle="success"
					bsSize="sm"
					style={{ marginTop: '5px' }}
					onClick={this.handleClick}
				>Evaluate</Button>
				<pre id="output">{this.state.result}</pre>
			</div>
		);
	}
}


// PROPERTY TYPES //

RShell.propTypes = {
	onResult: PropTypes.func,
	code: PropTypes.string,
	lines: PropTypes.number
};


// DEFAULT PROPERTIES //

RShell.defaultProps = {
	onResult() { },
	code: '',
	lines: 5
};


// EXPORTS //

export default RShell;
