// learning/rshell/index.js
// MODULES //

import React, { PropTypes } from 'react';
import ReactDom from 'react-dom';
import { Button } from 'react-bootstrap';
import request from 'request';
import ace from 'brace';
import 'brace/mode/r';
import 'brace/theme/katzenmilch';


// CONSTANTS //

const OPEN_CPU = 'https://public.opencpu.org';
const OPEN_CPU_POST = '/ocpu/library/base/R/identity/identity-d';
const GRAPHICS_REGEX = /graphics/;
const STDOUT_REGEX = /stdout/;


// R SHELL //

class RShell extends React.Component {
	constructor( props ) {
		super( props );

		this.state = {
			result: '',
			isGraphicOutput: false
		};

		this.handleClick = () => {
			this.setState({
				isGraphicOutput: false
			});
			const currentCode = this.editor.getValue();
			request.post( OPEN_CPU + OPEN_CPU_POST, {
				form: {
					x: currentCode
				}
			}, ( error, response, body ) => {
				const arr = body.split( '\n' );
				const res = arr[ 1 ];
				if ( !error ) {
					if ( GRAPHICS_REGEX.test( res ) === true ) {
						const imgURL = OPEN_CPU + res;
						this.setState({
							plot: imgURL,
							isGraphicOutput: true
						});
					}
					if ( STDOUT_REGEX.test( res ) === true ) {
						request.get( OPEN_CPU + res, ( err, getResponse, getBody ) => {
							this.setState({ result: getBody });
							this.props.onResult( this.state.result );
						});
					}
				}
			});
		};
	}

	componentDidMount() {
		this.editor = ace.edit( ReactDom.findDOMNode( this ).firstChild );
		this.editor.getSession().setMode( 'ace/mode/r' );
		this.editor.setTheme( 'ace/theme/katzenmilch' );
		this.editor.$blockScrolling = Infinity;
		this.editor.setValue( this.props.code );
		this.editor.setOptions({
			maxLines: this.props.lines,
			minLines: this.props.lines,
		});
		this.editor.getSession().setUseWrapMode( true );
	}
	render() {
		return (
			<div className="RShell" style={{ padding: '5px', border: '2px solid #73AD21' }}>
				<div id="ace"></div>
				<Button
					bsStyle="success"
					bsSize="sm"
					style={{ marginTop: '5px' }}
					onClick={this.handleClick}
				>Evaluate</Button>
				<div id="output">
					{this.state.isGraphicOutput ?
						<img role="presentation" src={this.state.plot}></img> :
						<pre id="output">{this.state.result}</pre>}
				</div>
			</div>
		);
	}
}


// PROPERTY TYPES //

RShell.propTypes = {
	onResult: PropTypes.func,
	code: PropTypes.string,
	lines: PropTypes.number
};


// DEFAULT PROPERTIES //

RShell.defaultProps = {
	onResult() { },
	code: '',
	lines: 5
};


// EXPORTS //

export default RShell;
