// general/select-input/index.js
// MODULES //

import React, { Component, PropTypes } from 'react';
import { ControlLabel, FormControl, FormGroup } from 'react-bootstrap';
import Input from 'general/input';


// SELECT INPUT //

class SelectInput extends Input {
	/**
	* Create a select input field.
	*
	* @param {Object} props - component properties (`onChange` callback and `defaultValue`)
	*/
	constructor( props ) {
		super( props );

		// Initialize state variables...
		this.state = {
			value: props.defaultValue
		};

		this.handleChange = ( event ) => {
			const value = event.target.value;
			this.setState({
				value
			});
			this.props.onChange( value );
		};
	}

	/*
	* React component render method.
	*/
	render() {
		return (
			<FormGroup controlId="formControlsSelect">
				<ControlLabel>{this.props.legend}</ControlLabel>
				<FormControl
					value={this.state.value}
					defaultValue={this.props.defaultValue}
					componentClass="select"
					placeholder="select"
					onChange={this.handleChange}
				>
					{this.props.options.map( e => {
						return <option value={e}>{e}</option>;
					})}
				</FormControl>
			</FormGroup>
		);
	}
};


// DEFAULT PROPERTIES //

SelectInput.defaultProps = {
	onChange() {},
	text: '',
	legend: '',
	options: []
};


// PROPERTY TYPES //

SelectInput.propTypes = {
	onChange: PropTypes.func,
	defaultValue: PropTypes.string,
	legend: PropTypes.string,
	options: PropTypes.array
};


// EXPORTS //

export default SelectInput;
