// general/check-box/index.js
// MODULES //

import React, { Component, PropTypes } from 'react';
import styles from 'styles';


// CHECK BOX //

class CheckBox extends Component {

	/**
	* Create a CheckBox instance
	*/
	constructor( props ) {
		super( props );

		this.state = {
			isChecked: false
		};

		/**
		* Event handler invoked once the checkbox is clicked by the user. Changes the
		* `isChecked` property and then invokes the user-supplied `onChange` callback function.
		*/
		this.handleChange = () => {
			this.setState({
				isChecked: !this.state.isChecked
			}, function onDone() {
				this.props.onChange( this.props.id, this.state.isChecked );
			});
		};
	}

	/**
	* React component render method
	*/
	render() {
		return (
			<div className="checkBox">
				<input
					type="checkbox"
					checked={this.state.isChecked}
					id={this.props.id}
					style={styles.feedbackCheckBox}
					onClick={this.handleChange}
				/>
				{this.props.text}
				<br />
			</div>
		);
	}
}


// DEFAULT PROPERTIES //

CheckBox.defaultProps = {
	onChange() {}
};


// PROPERTY TYPES //

CheckBox.propTypes = {
	onChange: PropTypes.func,
	id: PropTypes.string,
	text: PropTypes.string
};


// EXPORTS //

export default CheckBox;
