// general/number-input/index.js
// MODULES //

import React, { Component, PropTypes } from 'react';


// NUMBER INPUT //

class NumberInput extends Component {

	constructor() {
		super();
		this.state = {
			value: ''
		};

		this.handleChange = ( event ) => {
			this.setState({
				value: event.target.value
			});
		}
	}

	componentDidMount() {
		this.setState({
			value:  this.props.defaultValue
		})
	}

	render() {
		return (
			<div style={{
				marginTop: '12px',
				marginBottom: '12px'
			}}>
				<span style={{
					marginLeft: '8px'
				}}>{this.props.legend}</span>
				<input
					type="text"
					name="input"
					value={this.state.value}
					defaultValue={this.props.value}
					style={{
						marginRight: '8px',
						paddingLeft: '4px',
						paddingRight: '4px',
						background: 'gold',
						border: 'none',
						width: '80px',
						textAlign: 'center',
						float: 'right'
					}}
					onChange={this.handleChange}
				/>
			</div>
		);
	}
}


// DEFAULT PROPERTIES //

NumberInput.defaultProps = {
	legend: '',
	defaultValue: '0'
};


// PROPERTY TYPES //

NumberInput.propTypes = {
	legend: PropTypes.string,
	defaultValue: PropTypes.string
};


// EXPORTS //

export default NumberInput;
