// learning/tex/index.js
// MODULES //

import Radium from 'radium';
import React, { Component } from 'react';
import katex from 'katex';
import isString from 'validate.io-string';


// LATEX //

class TeX extends Component {
	/**
	* React component render method.
	*/
	render() {
		const input = isString( this.props.raw ) ? this.props.raw : this.props.raw.toString();
		let str;

		try {
			str = katex.renderToString( input, {
				displayMode: this.props.displayMode
			});
		} catch ( e ) {
			str = '';
		}
		let math = {
			__html: str
		};
		const inlineStyle = this.props.displayMode ?
			{} :
			{ whiteSpace: 'nowrap', display: 'inline-block' };
		return (
			<div
				className="tex"
				style={[ this.props.style, inlineStyle ]}
				onClick={this.props.onClick}
			>
				{this.props.displayMode ?
					<div
						style={{ float: 'right', marginTop: 5, marginRight: 5 }}
					>
					[{this.props.id}]
					</div> :
					<span />}
				<span
					ref="katex"
					dangerouslySetInnerHTML={math}
					aria-hidden={!!math}
					style={this.props.displayMode ? {} : { whiteSpace: 'nowrap' }}
				/>
			</div>
		);
	}

}

// PROPERTY TYPES //

TeX.propTypes = {
	raw: React.PropTypes.oneOfType([
		React.PropTypes.string,
		React.PropTypes.number
	]).isRequired,
	style: React.PropTypes.object,
	displayMode: React.PropTypes.bool,
	id: React.PropTypes.number,
	onClick: React.PropTypes.func
};


// DEFAULT PROPERTIES //

TeX.defaultProps = {
	onClick: null,
	displayMode: false,
	id: NaN,
	style: {}
};


// EXPORTS //

export default Radium( TeX );
