// general/text-area/index.js
// MODULES //

import React, { Component, PropTypes } from 'react';
import styles from 'styles';


// TEXT AREA //

class TextArea extends Component {
	/**
	* Create a text area
	*
	* @param {Object} props - component properties (`onChange` callback and `text`)
	*/
	constructor( props ) {
		super( props );

		// Initialize state variables...
		this.state = {
			text: props.text
		};

		/*
		* Event handler invoked when text area value changes. Updates `text` and invokes
		* `onChange` callback with the new text as its first argument
		*/
		this.handleChange = ( event ) => {
			const text = event.target.value;
			this.setState({ text });
			this.props.onChange( text );
		};
	}

	/*
	* React component render method.
	*/
	render() {
		return (
			<textarea
				style={styles.feedbackTextArea}
				value={this.state.text}
				onChange={this.handleChange}
			>
			</textarea>
		);
	}
}


// DEFAULT PROPERTIES //

TextArea.defaultProps = {
	onChange() {},
	text: ''
};


// PROPERTY TYPES //

TextArea.propTypes = {
	onChange: PropTypes.func,
	text: PropTypes.string
};


// EXPORTS //

export default TextArea;
