// general/checkbox-input/index.js
// MODULES //

import React, { Component, PropTypes } from 'react';
import Input from 'general/input';


// CHECKBOX INPUT //

class CheckboxInput extends Input {

	constructor( props ) {
		super( props );

		/**
		* Event handler invoked once the checkbox is clicked by the user. Changes the
		* `isChecked` property and then invokes the user-supplied `onChange` callback function.
		*/
		this.handleChange = ( event ) => {
			this.setState({
				value: event.target.checked
			}, () => {
				this.props.onChange( this.state.value );
				if ( this.context.autoUpdate ) {
					this.context.triggerDashboardClick();
				}
			});
		};
	}

	render() {
		return (
			<div style={{
				marginTop: '8px',
				marginLeft: '8px'
			}}>
				<input
					type="checkbox"
					checked={this.state.value}
					value="checkbox"
					onChange={this.handleChange}
					style={{
						verticalAlign: 'bottom',
						width: '24px',
						height: '24px'
					}}
				></input>
				<span
					style={{
						marginLeft: '12px'
					}}
				>{this.props.legend}</span>
			</div>
		);
	}
}


// DEFAULT PROPERTIES //

CheckboxInput.defaultProps = {
	onChange() {},
	defaultValue: false
};


// PROPERTY TYPES //

CheckboxInput.propTypes = {
	onChange: PropTypes.func,
	defaultValue: PropTypes.bool
};


// CONTEXT TYPES //

CheckboxInput.contextTypes = {
	triggerDashboardClick: PropTypes.func,
	autoUpdate: PropTypes.bool
};


// EXPORTS //

export default CheckboxInput;
