// general/slider-input/index.js
// MODULES //

import React, { Component, PropTypes } from 'react';
import Input from 'general/input';


// SLIDER INPUT //

class SliderInput extends Input {

	render() {
		return (
			<div>
				<input
					type="range"
					min={this.props.min}
					max={this.props.max}
					step={this.props.step}
					style={{
						marginLeft: '8px',
						marginTop: '8px',
						width: '70%',
						float: 'left'
					}}
					value={this.state.value}
					onChange={this.handleChange}
				/>
				<input
					type="number"
					name="input"
					style={{
						marginTop: '8px',
						marginRight: '8px',
						paddingLeft: '16px',
						paddingRight: '4px',
						background: 'gold',
						border: 'solid 1px darkgrey',
						textAlign: 'center',
						width: '80px',
						float: 'right'
					}}
					min={this.props.min}
					max={this.props.max}
					step={this.props.step}
					defaultValue={this.props.value}
					value={this.state.value}
					onChange={this.handleChange}
				/>
				<span style={{
					marginLeft: '8px',
					width: '90%',
					display: 'inline-block',
					float: 'none'
				}}>{this.props.legend}</span>
			</div>
		);
	}
}


// DEFAULT PROPERTIES //

SliderInput.defaultProps = {
	min: 0,
	max: 100,
	step: 1,
	defaultValue: 10
};


// PROPERTY TYPES //

SliderInput.propTypes = {
	min: PropTypes.number,
	max: PropTypes.number,
	step: PropTypes.number,
	defaultValue: PropTypes.number
};


// EXPORTS //

export default SliderInput;
