// learning/multiple-choice-question/index.js
// MODULES //

import React, { Component, PropTypes } from 'react';
import { Button, ListGroup, ListGroupItem, OverlayTrigger, Panel, Popover } from 'react-bootstrap';
import isArray from '@stdlib/utils/is-array';
import contains from '@stdlib/utils/contains';


// FUNCTIONS //

const Question = ( props ) =>
	<span className="question">
		<h3>{props.content}</h3>
		<span style={{ fontSize: '18px' }}>{props.task}:</span>
	</span>;

Question.PropTypes = {
	content: PropTypes.string.isRequired,
	task: PropTypes.string.isRequired
};

const AnswerOption = ( props ) => {
	let bsStyle;
	if ( props.correct === true ) {
		bsStyle = 'success';
	}
	else if ( props.correct === false ) {
		bsStyle = 'danger';
	}
	else if ( props.solution === true ) {
		// Case: User did not pick correct answer...
		bsStyle = 'warning';
	}

	const popover = <Popover id={props.answerContent}>
		<strong>{ props.solution ? 'Correct answer: ' : 'Incorrect answer: ' }</strong>{props.answerExplanation}
	</Popover>;

	if ( props.submitted ) {
		return (
			<OverlayTrigger
				trigger={[ 'click', 'hover' ]}
				placement="right"
				overlay={ props.answerExplanation ? popover : <span /> }
			>
				<ListGroupItem
					onClick={props.onAnswerSelected}
					bsStyle={bsStyle}
				>
					{props.answerContent}
				</ListGroupItem>
			</OverlayTrigger>
		);
	}
	else {
		return (
			<ListGroupItem
				onClick={props.onAnswerSelected}
				active={props.active}
			>
				{props.answerContent}
			</ListGroupItem>
		);
	}
};

AnswerOption.propTypes = {
	answerContent: PropTypes.string.isRequired,
	active: PropTypes.bool.isRequired,
	answerExplanation: PropTypes.string,
	onAnswerSelected: PropTypes.func.isRequired
};


// MULTIPLE CHOICE QUESTION //

class MultipleChoiceQuestion extends Component {

	constructor( props ) {
		super( props );

		let active = isArray( this.props.solution ) ?
			new Array( props.answers.length ) :
			null;

		this.state = {
			submitted: false,
			active,
			correct: new Array( props.answers.length ),
		};

		this.submitQuestion = () => {
			let sol = this.props.solution;
			let newCorrect = new Array( this.props.answers.length );

			if ( isArray( sol ) ) {
				for ( let i = 0; i < this.state.active.length; i++ ) {
					if ( this.state.active[ i ] === true ) {
						if ( contains( sol, i ) ) {
							newCorrect[ i ] = true;
						} else {
							newCorrect[ i ] = false;
						}
					}
				}
				let active = new Array( props.answers.length );
				this.setState({
					correct: newCorrect,
					submitted: true,
					active
				});
			}
			else {
				for ( let i = 0; i < newCorrect.length; i++ ) {
					if ( this.state.active === i ) {
						if ( i === sol ) {
							newCorrect[ i ] = true;
						} else {
							newCorrect[ i ] = false;
						}
					}
				}
				let active = null;
				this.setState({
					correct: newCorrect,
					submitted: true,
					active
				});
			}
		};
	}

	render() {
		const props = this.props;
		const allowMultipleAnswers = isArray( this.props.solution );

		const renderAnswerOptionsMultiple = ( key, id ) => {
			let isSolution = contains( this.props.solution, id );
			return (
				<AnswerOption
					key={key.content}
					answerContent={key.content}
					answerExplanation={key.explanation}
					active={this.state.active[ id ]}
					correct={this.state.correct[ id ]}
					submitted={this.state.submitted}
					solution={isSolution}
					onAnswerSelected={ () => {
						if ( !this.state.submitted ) {
							let newActive = this.state.active.slice();
							newActive[ id ] = !newActive[ id ];
							this.setState({
								active: newActive
							});
						}
					}}
				/>
			);
		};

		const renderAnswerOptionsSingle = ( key, id ) => {
			let isSolution = this.props.solution === id;
			return (
				<AnswerOption
					key={key.content}
					answerContent={key.content}
					answerExplanation={key.explanation}
					active={this.state.active === id}
					correct={this.state.correct[ id ]}
					submitted={this.state.submitted}
					solution={isSolution}
					onAnswerSelected={ () => {
						if ( !this.state.submitted ) {
							this.setState({
								active: id
							});
						}
					}}
				/>
			);
		};

		return (
			<Panel className="multipleChoiceQuestion" style={{
				margin: '0 auto 10px',
				maxWidth: 600,
				marginTop: '8px'
			}}>
				<Question
					content={props.question}
					task={ allowMultipleAnswers ? 'Choose all that apply' : 'Pick the correct answer' }
				/>
				<ListGroup fill>
					{ allowMultipleAnswers ?
						props.answers.map( renderAnswerOptionsMultiple ) :
						props.answers.map( renderAnswerOptionsSingle )
					}
				</ListGroup>
				<Button
					bsSize="small"
					bsStyle="success"
					block fill
					onClick={this.submitQuestion}
					disabled={this.state.submitted}
				>Submit</Button>
			</Panel>
		);
	}
}


// DEFAULT PROPERTIES //

MultipleChoiceQuestion.propTypes = {
	answers: []
};


// PROPERTY TYPES //

MultipleChoiceQuestion.propTypes = {
	solution: PropTypes.oneOfType([
		PropTypes.number,
		PropTypes.array
	]).isRequired,
	answers: PropTypes.array.isRequired,
};


// EXPORTS //

export default MultipleChoiceQuestion;
