// learning/feedback/feedback-form/index.js
// MODULES //

import radium from 'radium';
import React, { Component, PropTypes } from 'react';
import styles from 'styles';

import TextArea from 'general/text-area';
import CheckBox from 'general/check-box';


// FEEDBACK FORM //

class FeedbackForm extends Component {
	/*
	* Create a feedback form to collect detailed user feedback.
	*/
	constructor( props ) {
		super( props );

		// Set initial states for feedback form...
		this.state = {
			comment: '',
			checked: [
				false,
				false,
				false,
				false
			]
		};
		/**
		* Event handler invoked when the value of the text area changes.
		*
		* @param {String} text - current content of the text area
		*
		* @returns {Void} updates the `comment` state of the component
		*/
		this.handleTextAreaChange = ( text ) => {
			this.setState({ comment: text });
		};
		/**
		* Event handler invoked when one of the check boxes is clicked. Updates the `checked`
		* state variable of the component.
		*
		* @param {Number} id - id of the checkbox
		* @param {Boolean} checked - checked property value of the check box
		*/
		this.handleCheckBoxChange = ( id, checked ) => {
			const arr = this.state.checked.slice();
			arr[ id ] = checked;
			this.setState({
				checked: arr
			});
		};
	}

	/*
	* React component render method.
	*/
	render() {
		return (
			<div
				className="feedbackForm"
				style={[ styles.feedbackForm, { top: this.props.top } ]}
			>
				<button
					ref="closeButton"
					style={styles.closeButton}
					onClick={this.props.closeHandler}
				>Close</button>
				<div style={styles.feedbackTitle}>Feedback</div>
				<form style={styles.feedbackFormElem}>
					<CheckBox
						id="0"
						text="I do not understand this at all."
						onChange={this.handleCheckBoxChange}
					/>
					<CheckBox
						id="1"
						text="This needs a more detailed explanation."
						onChange={this.handleCheckBoxChange}
					/>
					<CheckBox
						id="2"
						text="I can't follow the logic."
						onChange={this.handleCheckBoxChange}
					/>
					<CheckBox
						id="3"
						text="I have the following comments:"
						onChange={this.handleCheckBoxChange}
					/>
					<TextArea text={this.state.comment} onChange={this.handleTextAreaChange} />
					<input
						ref="submitButton"
						type="submit"
						style={styles.submitButton}
						onClick={this.props.submitHandler}
						value="Submit"
					/>
				</form>
			</div>
		);
	}
}


// DEFAULT PROPERTIES //

FeedbackForm.defaultProps = {
	closeHandler() {},
	submitHandler() {},
	top: 0
};


// PROPERTY TYPES //

FeedbackForm.propTypes = {
	closeHandler: PropTypes.func,
	submitHandler: PropTypes.func,
	top: PropTypes.number
};


// EXPORTS //

export default radium( FeedbackForm );
