// learning/video-player/index.js
// MODULES //

import React, { Component, PropTypes } from 'react';
import ReactPlayer from 'react-player';
import Dimensions from 'react-dimensions';


// FUNCTIONS //

function calculateMargin( containerWidth, targetWidth ) {
	const x = ( containerWidth - targetWidth ) / 2.0;
	return `${x | 0}px`;
}

// VIDEO //

class Video extends Component {

	render() {
		return (
			<div
				style={{
					position: 'relative',
					marginLeft: calculateMargin( this.props.containerWidth, this.props.width ),
					marginRight: calculateMargin( this.props.containerWidth, this.props.width ),
					marginTop: '30px'
				}}
			>
				<ReactPlayer {...this.props} />
			</div>
		);
	}

}


// PROPERTY TYPES //

Video.propTypes = {
	url: PropTypes.string,
	controls: PropTypes.bool,
	playing: PropTypes.bool,
	volume: PropTypes.number,
	height: PropTypes.number,
	width: PropTypes.number
};


// DEFAULT PROPERTIES //

Video.defaultProps = {
	url: '',
	controls: false,
	playing: false,
	volume: 0.8,
	height: 360,
	width: 640
};


// EXPORTS //

export default Dimensions({
	elementResize: true
})( Video );
