// general/dashboard/index.js
// MODULES //

import React, { Component, PropTypes } from 'react';
import { Button } from 'react-bootstrap';
import Input from 'general/input';
import NumberInput from 'general/number-input';
import SliderInput from 'general/slider-input';
import { findAllChildren } from 'utils/find-nodes';


// DASHBOARD //

class Dashboard extends Component {

	constructor() {
		super();

		this.handleClick = () => {
			let inputs = findAllChildren( this, Input );
			let args = inputs.map( v => {
				if (
					v instanceof NumberInput ||
					v instanceof SliderInput
				) {
					return parseFloat( v.state.value );
				}
				else {
					return v.state.value;
				}
			});
			this.props.onGenerate( ...args );

			const lesson = window.lesson || global.lesson;
			if ( lesson && !lesson.unmounted ) {
				lesson.forceUpdate();
			}
		};
	}

	componentDidMount(){
		this.handleClick();
	}

	render() {
		return (
			<div className="dashboard" style={{
				margin: '5px',
				padding: '5px',
				border: '2px solid lightblue',
				background: 'white',
				maxWidth: '600px'
			}}>
				<label>{this.props.title}</label>
				<p>{this.props.description}</p>
				<hr
					style={{
						marginTop: 0,
						marginBottom: 8
					}}
				/>
				{this.props.children}
				<hr
					style={{
						marginTop: 0,
						marginBottom: 8
					}}
				/>
				<Button
					bsStyle="info"
					onClick={this.handleClick}
				>Generate</Button>
			</div>
		);
	}
}


// DEFAULT PROPERTIES //

Dashboard.defaultProps = {
	onGenerate: function(){}
};


// PROPERTY TYPES //

Dashboard.propTypes = {
	onGenerate: PropTypes.func,
	description: PropTypes.string,
	title: PropTypes.string
};


// EXPORTS //

export default Dashboard;
