// general/dashboard/index.js
// MODULES //

import React, { Component, PropTypes } from 'react';
import { Button, Panel } from 'react-bootstrap';
import Input from 'general/input';
import NumberInput from 'general/number-input';
import SliderInput from 'general/slider-input';
import { findAllChildren } from 'utils/find-nodes';


// DASHBOARD //

class Dashboard extends Component {

	constructor() {
		super();

		this.handleClick = () => {
			let inputs = findAllChildren( this, Input );
			let args = inputs.map( v => {
				if (
					v instanceof NumberInput ||
					v instanceof SliderInput
				) {
					return parseFloat( v.state.value );
				}
				else {
					return v.state.value;
				}
			});
			if ( this.props.id ) {
				global.ISLE.session.log({
					id: this.props.id,
					type: 'DASHBOARD_CLICK_GENERATE',
					value: args
				});
			}
			this.props.onGenerate( ...args );

			const lesson = window.lesson || global.lesson;
			if ( lesson && !lesson.unmounted ) {
				lesson.forceUpdate();
			}
		};
	}

	componentDidMount(){
		this.handleClick();
	}

	render() {
		return (
			<Panel
				className="dashboard"
				bsStyle="info"
				header={<h4>{this.props.title}</h4>}
				style={{
					marginTop: '5px',
					marginBottom: '5px',
					background: 'white',
					maxWidth: '600px'
				}}
			>
				<p>{this.props.description}</p>
				{this.props.children}
				<Button
					bsStyle="info"
					onClick={this.handleClick}
					style={{
						marginTop: '5px',
						marginBottom: '5px',
						top: '-10px'
					}}
					block
				>Generate</Button>
			</Panel>
		);
	}
}


// DEFAULT PROPERTIES //

Dashboard.defaultProps = {
	onGenerate: function(){}
};


// PROPERTY TYPES //

Dashboard.propTypes = {
	onGenerate: PropTypes.func,
	description: PropTypes.string,
	title: PropTypes.string
};


// EXPORTS //

export default Dashboard;
