// learning/r/output/index.js
// MODULES //

import React, { Component, PropTypes } from 'react';
import Dimensions from 'general/dimensions';
import { Table, Column, Cell } from 'fixed-data-table';
import { Button, Modal } from 'react-bootstrap';
import DOMPurify from 'dompurify';
import request from 'request';
import Spinner from 'general/spinner';
import floor from '@stdlib/math/base/special/floor';
import isArray from '@stdlib/utils/is-array';


// CONSTANTS //

const OPEN_CPU_DEFAULT = 'http://phd-serv5.heinz.cmu.edu';
const OPEN_CPU_POST = '/ocpu/library/base/R/identity/identity-d';
const STDOUT_REGEX = /stdout/;


// FUNCTIONS //

const showResult = ( res ) => {
	if ( res ) {
		let sanitized = {
			__html: DOMPurify.sanitize( res )
		};
		return <pre id="output"><span dangerouslySetInnerHTML={sanitized} /></pre>;
	} else {
		return <span />;
	}
};

// R OUTPUT //

class ROutput extends Component {

	constructor( props ) {
		super( props );

		this.state = {
			result: null,
			last: ''
		};

		this.getResult = ( nextProps ) => {
			console.log( nextProps )
			let code;
			if ( nextProps ) {
				code = nextProps.code;
			} else {
				code = this.props.code;
			}
			if ( code !== this.state.last ) {
				this.setState({
					last: this.props.code,
				});

				const OPEN_CPU = global.ISLE.rshell && global.ISLE.rshell.server ?
					global.ISLE.rshell.server :
					OPEN_CPU_DEFAULT;

				let fullCode = '';
				let prependCode = isArray( this.props.prependCode ) ?
					this.props.prependCode.join( '\n' ) :
					this.props.prependCode;
				prependCode += '\n';

				fullCode = prependCode + code;

				request.post( OPEN_CPU + OPEN_CPU_POST, {
					form: {
						x: fullCode
					}
				}, ( error, response, body ) => {
					console.log( body )
					if ( !error ) {
						const arr = body.split( '\n' );
						for ( let i = 0; i < arr.length; i++ ) {
							let elem = arr[ i ];
							console.log( elem )
							if ( STDOUT_REGEX.test( elem ) === true ) {
								request.get( OPEN_CPU + elem, ( err, getResponse, getBody ) => {
									this.setState({
										result: getBody
									});
								});
								break;
							}
						}
					}
				});
			}
		};
	}

	componentWillReceiveProps( props ) {
		this.getResult( props );
	}

	render() {
		return (
			<span className="ROutput">
				{ this.state.result ?
					<div
						style={{
							marginLeft: 'auto',
							marginRight: 'auto',
							marginTop: '10px',
							marginBottom: '10px'
						}}
					>
						{ showResult( this.state.result ) }
					</div> :
					<span />
				}
			</span>
		);
	}
}


// PROPERTY TYPES //

ROutput.propTypes = {
	code: PropTypes.string
};


// DEFAULT PROPERTIES //

ROutput.defaultProps = {
	code: ''
};


// EXPORTS //

export default ROutput;
