// general/lesson-submit/index.js
// MODULES //

import React, { Component, PropTypes } from 'react';
import { Button, Modal, ControlLabel, FormControl, FormGroup } from 'react-bootstrap';
import request from 'request';


// LESSON SUBMIT //

class LessonSubmit extends Component {

	constructor( props ) {
		super( props );

		this.state = {
			showUserModal: false,
			user: '',
			email: ''
		};

		this.closeUserModal = () => {
			this.setState({
				showUserModal: false
			});
		};

		this.handleClick = () => {
			const str = 'ISLE_USER_' + ISLE.server;
			let item = localStorage.getItem( str );
			if ( !item ) {
				this.setState({
					showUserModal: true
				});
			}
			else {
				global.session.finalize();
				global.lesson.addNotification({
					title: 'Completed',
					message: 'Lesson successfully completed. You will receive a confirmation email shortly.',
					level: 'success',
					position: 'tr'
				});
				const user = JSON.parse( item );
				if ( this.props.mail ) {
					global.sendMail( this.props.mail, user.email );
				}
			}
		};

		this.handleEmailInput = ( e ) => {
			this.setState({ email: e.target.value });
		};

		this.handleUserInput = ( e ) => {
			this.setState({ user: e.target.value });
		};

		this.submitRegistration = () => {
			request.get({
				url: global.ISLE.server + '/user',
				qs: {
					email: this.state.email,
					name: this.state.user
				}
			}, ( error, response, body ) => {
				const str = 'ISLE_USER_' + ISLE.server;
				localStorage.setItem( str, body );
			});
		};
	}

	render() {
		return (
			<div className="well" style={{
				maxWidth: 400,
				margin: '0 auto 10px',
				marginTop: '30px',
				fontFamily: 'Arial'
			}}>
				<Button bsStyle="primary" bsSize="large" onClick={this.handleClick} block>{this.props.label}</Button>
				<Modal show={this.state.showUserModal} onHide={this.closeUserModal}>
					<Modal.Header closeButton>
						<Modal.Title>User Registration</Modal.Title>
					</Modal.Header>
					<Modal.Body>
						<FormGroup>
							<ControlLabel>Email Address</ControlLabel>
							<FormControl
								type="text"
								value={this.state.email}
								placeholder="Enter text"
								onChange={this.handleEmailInput}
							/>
							<ControlLabel>Full Name</ControlLabel>
							<FormControl
								type="text"
								value={this.state.user}
								placeholder="Enter text"
								onChange={this.handleUserInput}
							/>
						</FormGroup>
					</Modal.Body>
					<Modal.Footer>
						<Button onClick={this.closeUserModal}>Close</Button>
						<Button bsStyle="primary" onClick={this.submitRegistration}>Submit</Button>
					</Modal.Footer>
				</Modal>
			</div>
		);
	}
}


// DEFAULT PROPERTIES //

LessonSubmit.defaultProps = {
	label: 'Finish lesson',
	mail: null
};


// PROPERTY TYPES //

LessonSubmit.propTypes = {
	label: PropTypes.string,
	mail: PropTypes.string
};


// EXPORTS //

export default LessonSubmit;
