// general/text-input/index.js
// MODULES //

import React, { Component, PropTypes } from 'react';
import Input from 'general/input';


// TEXT INPUT //

class TextInput extends Input {

	constructor( props ) {
		super( props );

		this.state = {
			value: props.defaultValue
		};

		this.handleChange = ( event ) => {
			const value = event.target.value;
			this.setState({
				value
			}, () => {
				this.props.onChange( value );
				if ( this.context.autoUpdate ) {
					this.context.triggerDashboardClick();
				}
			});

		};
	}

	render() {
		return (
			<div style={{
				marginBottom: '4px',
				marginTop: '4px',
				clear: 'both'
			}}>
				<span style={{
					marginLeft: '8px'
				}}>
					<label>{this.props.legend}:</label>
					{ this.props.description ?
						<span> {this.props.description}</span> :
						<span />
					}
				</span>
				<input
					type="text"
					name="input"
					value={this.state.value}
					style={{
						marginRight: '8px',
						paddingLeft: '16px',
						paddingRight: '4px',
						border: 'solid 1px darkgrey',
						background: 'gold',
						width: this.props.width,
						textAlign: 'left',
						float: 'right'
					}}
					onChange={this.handleChange}
				/>
			</div>
		);
	}
}


// DEFAULT PROPERTIES //

TextInput.defaultProps = {
	defaultValue: '',
	width: 80,
	onChange(){}
};


// PROPERTY TYPES //

TextInput.propTypes = {
	defaultValue: PropTypes.string,
	width: PropTypes.number,
	onChange: PropTypes.func
};


// CONTEXT TYPES //

TextInput.contextTypes = {
	triggerDashboardClick: PropTypes.func,
	autoUpdate: PropTypes.bool
};


// EXPORTS //

export default TextInput;
